core/src/test/java/io/questdb/test/cairo/meta/BlockFileTest.java

```diff
@@ -0,0 +1,737 @@
+                try {
+                    readAllBlocks(path, 1);
+                    Assert.fail("checksum mismatch expected");
+                } catch (Exception e) {
+                    Assert.assertTrue(e.getMessage().contains("meta file checksum mismatch"));
+                }
+            }
+        });
+    }
+
+    @Test
+    public void testReadWriteConcurrently() throws Exception {
+        assertMemoryLeak(() -> {
+            FilesFacade ff = configuration.getFilesFacade();
+            Rnd rnd = TestUtils.generateRandom(LOG);
+            int readerThreads = 4;
+            CyclicBarrier start = new CyclicBarrier(readerThreads + 1);
+            AtomicInteger done = new AtomicInteger();
+            int iterations = 1000;
+            Thread writerThread = new Thread(() -> {
+                try (Path path = getDefinitionFilePath("test")) {
+                    start.await();
+                    for (int i = 0; i < iterations; i++) {
+                        int msg = rnd.nextInt(4);
+                        try (BlockFileWriter writer = new BlockFileWriter(ff, commitMode)) {
+                            writer.of(path.$());
+                            switch (msg) {
+                                case 0:
+                                    commitMsgAVersion1(writer.append());
+                                    break;
+                                case 1:
+                                    commitMsgAVersion2(writer.append());
+                                    break;
+                                case 2:
+                                    commitMsgBVersion1(writer.append());
+                                    break;
+                                case 3:
+                                    commitMsgCVersion1(writer.append());
+                                    break;
+                            }
+                            writer.commit();
+                        }
+                    }
+                } catch (Exception e) {
+                    LOG.error().$("Error in writer thread: ").$(e).$();
+                } finally {
+                    done.incrementAndGet();
+                }
+            });
+
+            Thread[] readers = new Thread[readerThreads];
+            for (int th = 0; th < readerThreads; th++) {
+                Thread readerThread = new Thread(() -> {
+                    try (Path path = getDefinitionFilePath("test")) {
+                        start.await();
+                        for (int i = 0; i < iterations; i++) {
+                            Os.sleep(1); // interleave reads and writes
+                            readAllBlocks(path, 1);
+                        }
+                    } catch (Exception e) {
+                        LOG.error().$("Error in reader thread: ").$(e).$();
+                    }
+                });
+                readers[th] = readerThread;
+                readerThread.start();
+            }
+
+            writerThread.start();
```

Comment:
Please assert that the file size never goes beyond 2x max block size + headers.

Reply:
The actual condition is `< 3 * max_version_size`, as explained in the comments.